### PR TITLE
MSI Authentication (scope: VMs)

### PR DESCRIPTION
**Pull request by balajikris**
Basing this PR on top of Amar's AuthFromFile PR which is still open. 

Implementation of this PR is in the following commits: 
1. b1c952d -- implement MSI Token Credentials and add login methods
1. 0cb92a0 -- tests

### CODE REVIEW

**Comment by lostintangent** [comment]
This looks awesome! In what environments can this new login method be used? Just VMs that we're provisioned with a managed identity?

**Comment by balajikris** [comment]
Hey @lostintangent -- thanks! Currently, we're looking to bring this on for VMs that have been provisioned with the MSI extension. @yugangw-msft is my best guess as to someone who _may_ have a better idea about how or when we'll onboard other services with MSI ?

File: package.json [hunk 1]
```diff
@@ -1,157 +1,157 @@
...
+    "grunt": "~0.4",
+    "grunt-contrib-connect": "^0.10.1",
+    "grunt-contrib-symlink": "^0.3.0",
+    "grunt-devserver": "~0.6",
+    "grunt-jsdoc": "~0.6",
+    "grunt-gh-pages": "1.1.0",
+    "jshint": "2.9.4",
+    "minami": "git://github.com/devigned/minami#master",
+    "mocha": "^3.2.0",
+    "nock": "^9.0.14",
```
Comment (amarzavery) [inline package.json]: dont change the nock version for rollup azure.
Instead add  nock as a dev dependency in ms-rest-azure's package.json and this test over there. Since this is more of a unit test and not a scenario/live test anymore.

File: runtime/ms-rest-azure/lib/credentials/msiTokenCredentials.js [hunk 2]
```diff
@@ -12,25 +12,25 @@ class MSITokenCredentials {
       throw new Error('domain must be a non empty string.');
```
Comment (amarzavery) [inline runtime/ms-rest-azure/lib/credentials/msiTokenCredentials.js]: Please take a look at the flow implemented for MSI auth flow in CLI. 
- https://github.com/Azure/azure-cli/blob/master/src/azure-cli-core/azure/cli/core/_profile.py#L532-L592.

The CLI asks for the tenantId against which the token is required. We are not asking that from the customer.

File: runtime/ms-rest-azure/lib/login.js [hunk 3]
```diff
@@ -1,699 +1,732 @@
...
+    throw new Error('callback cannot be null or undefined.');
+  }
+  const creds = new MSITokenCredentials(domain, options)
+  creds.getToken(function (err) {
+    if (err) return callback(err);
+    return callback(null, creds);
+  });
+}
+
+exports.withMSI = function withMSI(domain, options, optionalCallback) {
```
Comment (amarzavery) [inline runtime/ms-rest-azure/lib/login.js]: missing documentation

Comment (amarzavery) [inline runtime/ms-rest-azure/lib/login.js]: export this in lib/msRestAzure.js

Comment (amarzavery) [inline runtime/ms-rest-azure/lib/login.js]: update the changelog

File: runtime/ms-rest-azure/lib/credentials/msiTokenCredentials.js (hunk 2 above)
Comment (balajikris) [inline runtime/ms-rest-azure/lib/credentials/msiTokenCredentials.js]: Isn't tenantID also called as Domain? I could have used the term `tenantID` but all the other login methods call it `domain`, so I stuck to it for consistency. Unless I'm completely off here.

File: runtime/ms-rest-azure/lib/login.js (hunk 3 above)
Comment (balajikris) [inline runtime/ms-rest-azure/lib/login.js]: work in progress.

File: runtime/ms-rest-azure/test/msiTokenCredentialTests.js [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+ */
+
+// all the requires
+var should = require('should');
+var nock = require('nock');
+var MSITokenCredential = require('../lib/credentials/msiTokenCredentials');
+
+
+describe('MSI Authentication', function () {
+    before(function (done) {
```
Comment (amarzavery) [inline runtime/ms-rest-azure/test/msiTokenCredentialTests.js]: indentation 2 spaces

File: runtime/ms-rest-azure/package.json [hunk 5]
```diff
@@ -28,7 +28,7 @@
     "async": "0.2.7",
     "uuid": "^3.0.1",
     "adal-node": "^0.1.17",
-    "ms-rest": "^2.2.2",
+    "ms-rest": "^2.2.1",
```
Comment (amarzavery) [inline runtime/ms-rest-azure/package.json]: why was this changed to 2.2.1? We intend to publish this with new version as the min dependency.

Comment (balajikris) [reply runtime/ms-rest-azure/package.json]: good catch -- had that to get tests passing, verified that CI passes (https://travis-ci.org/Azure/azure-sdk-for-node/jobs/270179285) .. Will revert this now.

**Comment by balajikris** [comment]
Status: All tests pass w/ the right version of ms-rest. See [here](https://travis-ci.org/Azure/azure-sdk-for-node/builds/270179284?utm_source=github_status&utm_medium=notification).
